### PR TITLE
Make UnicastHostsProvider extension pull based

### PR DESCRIPTION
**Pull request by rjernst**
This change moves providing UnicastHostsProvider for zen discovery to be
pull based, adding a getter in DiscoveryPlugin. A new setting is added,
discovery.zen.hosts_provider, to separate the discovery type from the
hosts provider for zen when it is selected. Unfortunately existing
plugins added ZenDiscovery with their own name in order to just provide
a hosts provider, so there are already many users setting the hosts
provider through discovery.type. This change also includes backcompat,
falling back to discovery.type when discovery.zen.hosts_provider is not
set.

### CODE REVIEW

File: plugins/discovery-ec2/src/main/java/org/elasticsearch/plugin/discovery/ec2/Ec2DiscoveryPlugin.java [hunk 1]
```diff
@@ -112,6 +101,15 @@ public void onModule(DiscoveryModule discoveryModule) {
     }
 
     @Override
+    public Map<String, Supplier<UnicastHostsProvider>> getZenHostsProviders(TransportService transportService,
```
Comment (dadoonet) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/plugin/discovery/ec2/Ec2DiscoveryPlugin.java]: <3

**Comment by s1monw** [comment]
LGTM thanks ryan

**Comment by bleskes** [comment]
@rjernst I relooked at the context of #21049 - Is there a follow up plan to document the host provider settings and update, for example, https://www.elastic.co/guide/en/elasticsearch/plugins/5.0/discovery-ec2-discovery.html ?

**Comment by rjernst** [comment]
@bleskes Good catch, I will make a followup PR to update documentation.

**Comment by rjernst** [comment]
@bleskes Added docs in #21382

**Pull request merged**